client/question/services/FeedbackGeneratorService.js

```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
-      // TODO(eyurko): Add varied statements for when code is incorrect.
-      feedback.appendTextParagraph('Your code produced the following result:');
-      if (outputFunctionName) {
+      // Check if new / next test suite
+      if (testSuiteId !== previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
+        if (correctnessTestStates.hasOwnProperty(testCaseKey)) {
+          feedback.appendTextParagraph(
+            'It looks like there was a regression in your code. Your code ' +
+            'worked for the following, but it now fails:');
+          feedback.appendCodeParagraph(
+            'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
+            'Expected Output: ' + _jsToHumanReadable(allowedOutputExample));
+          return feedback;
+        }
+        correctnessTestStates[testCaseKey] = CORRECTNESS_STATE_STARTING;
+      }
+      // If the suite ID corresponds to the sample input, the question will
+      // already have displayed the input and expected output, so we advance
+      // the correctness state to "expected output displayed".
+      if (testSuiteId === TEST_SUITE_ID_SAMPLE_INPUT) {
+        correctnessTestStates[testCaseKey] =
+          CORRECTNESS_STATE_EXPECTED_OUTPUT_DISPLAYED;
+      }
+      switch (correctnessTestStates[testCaseKey]) {
+      case CORRECTNESS_STATE_STARTING:
+        // Display input user should try
+        feedback.appendTextParagraph(
+          _getCorrectnessFeedbackString('INPUT_TO_TRY'));
         feedback.appendCodeParagraph(
-          'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          ('Output (after running ' + outputFunctionName + '): ' +
-            _jsToHumanReadable(observedOutput)));
+          'Input: ' + _jsToHumanReadable(testCase.getInput()));
+        correctnessTestStates[testCaseKey] =
+          CORRECTNESS_STATE_INPUT_DISPLAYED;
+        return feedback;
+      case CORRECTNESS_STATE_INPUT_DISPLAYED:
+        // Display expected output
         feedback.appendTextParagraph(
-          'However, the expected output of ' + outputFunctionName + ' is:');
-      } else {
+          _getCorrectnessFeedbackString('EXPECTED_OUTPUT'));
         feedback.appendCodeParagraph(
           'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          'Output: ' + _jsToHumanReadable(observedOutput));
-        feedback.appendTextParagraph('However, the expected output is:');
+          'Expected Output: ' +
+          _jsToHumanReadable(allowedOutputExample));
+        correctnessTestStates[testCaseKey] =
+          CORRECTNESS_STATE_EXPECTED_OUTPUT_DISPLAYED;
+        return feedback;
+      default:
+        // Allow user to display code output
+        feedback.appendTextParagraph(
+          _getCorrectnessFeedbackString('OUTPUT_ENABLED'));
+        feedback.appendOutputParagraph(
+          'Input: ' + _jsToHumanReadable(testCase.getInput()) +
```

Comment:
Just to check, did you also mean to show the input and expected output when the display is toggled? I thought at this point the input / expected output would always be shown.

Reply:
This is to provide context for the output displayed. When we implement hints, hints can be displayed before the option to display output; so, at that point, we want to provide the input and expected output as context.